### PR TITLE
Cache by header value: a new Header property in (File)CacheOptions configuration of a route

### PR DESCRIPTION
**Pull request by EngRajabi**
## New Feature 
Add cache in route with custom header. The header can be both full and empty.

Example:
```json
"FileCacheOptions": {
  "TtlSeconds": 15,
  "Region": "somename",
  "Header": "authorization"
}
```

### CODE REVIEW

**Comment by kooshan** [comment]
@TomPallister  Would you please merge this?

**Comment by EngRajabi** [comment]
fix Conflict

**Comment by EngRajabi** [comment]
@TomPallister

**Comment by HaydnDias** [comment]
Is there any update on this?

**Review by garybond — COMMENTED** [review]
Be nice to have this feature in the docs

**Review by EngRajabi — COMMENTED** [review]
ok

**Review by EngRajabi — COMMENTED** [review]
merge with master.
add to document.
please approve

**Comment by raman-m** [comment]
@EngRajabi  Hi Mohsen!
What issue is this PR related to? Number?

**Comment by raman-m** [comment]
@EngRajabi  Lazy bones!
[Come on!](#issuecomment-1554657265) 😉

**Comment by EngRajabi** [comment]
> @EngRajabi Lazy bones! [Come on!](#issuecomment-1554657265) 😉
> 
> Why not to resolve all merge conflicts, please?

I didn't expect this repo to be active after a few years. I was surprised
I will solve the PR problems by Friday. Thankful

**Comment by raman-m** [comment]
image

Your develop branch is not default! Make it default please!

---

image

Your develop branch has [redundant commits](../compare/develop...EngRajabi:Ocelot:develop) and it is not synced to head repo (ThreeMammals:develop)!
It should be updated and the same as head repo!

### To fix that
Just delete **develop** branch using GitHub UI and make `git fetch --all`, and remove develop locally.
After that use this script to pull develop from head (upstream) repository:
```shell
git remote add upstream https://github.com/ThreeMammals/Ocelot.git // add original repo as upstream remote
git fetch upstream                                                 // pull changes from original repo
git checkout -b develop upstream/develop                           // pull down develop branch locally
git push -u origin develop                                         // push develop branch to fork
```
Don't forget to make **develop** branch as default one.

Let me know upgrading results for **develop** branch, please!
Thank you!

**Review by EngRajabi — COMMENTED** [review]
merge conflict
review done

**Comment by raman-m** [comment]
Why did you [merge develop in invalid state to feature branch](#commits-pushed-ff1dbb6)?
You had to postpone such merging.
And you had to [use my instructions to recover develop branch](#issuecomment-1640049147) to canonic valid state, but you've ignored them! 😢 

Now there is no conflicts which is good news, but we have a problem with pipeline builds! 🥲

**Comment by raman-m** [comment]
Hi Mohsen!
I've done with [rebasing your feature](#commits-pushed-3983da8) branch! And we have [green valid build](https://circleci.com/gh/ThreeMammals/Ocelot/1266) now. 🎉
Your changes look very nice now, and it is easy to read them.

So, we can start code review process now... 😉

**Comment by EngRajabi** [comment]
> Hi Mohsen! I've done with [rebasing your feature](#commits-pushed-3983da8) branch! That was not easy... 😥 And we have [green valid build](https://circleci.com/gh/ThreeMammals/Ocelot/1266) now. 🎉 Your changes look very nice now, and it is easy to read them.
> 
> So, we can start code review process now... 😉
> 
> Is this PR related to some [issue](../issues)?

Thanks
I don't have an open issue with this.
But there is a need to be able to set the cache specifically for each user.
The default mode of the cache is for all users because there are not many APIs that return the same thing for all users.

**Review by EngRajabi — COMMENTED** [review]
review done

File: src/Ocelot/Cache/CacheKeyGenerator.cs
Comment (raman-m) [inline src/Ocelot/Cache/CacheKeyGenerator.cs:29]: Because of this significant changes:
- Do we have unit tests for the `GenerateRequestCacheKey` method?
- Could you add unit tests to cover this new logic/lines please?

**Review by raman-m — CHANGES_REQUESTED** [review]
Need more unit tests!

**Review by RaynaldM — DISMISSED** [review]

File: docs/features/caching.rst
Comment (wast) [inline docs/features/caching.rst:30]: What is HEader ?

File: src/Ocelot/Cache/Middleware/OutputCacheMiddleware.cs
Comment (wast) [inline src/Ocelot/Cache/Middleware/OutputCacheMiddleware.cs:45]: Why is there mixed naming with downstream and downStream?

Comment (raman-m) [reply src/Ocelot/Cache/Middleware/OutputCacheMiddleware.cs:45]: @wast 
Thanks for the review!
I agree it should be `downstreamRequestCacheKey`

Another convention issue is `string downstreamRequestCacheKey`, but it should be `var downstreamRequestCacheKey`.
This is my fault. Maybe during resolving merge conflicts I've skipped the issue.

**Review by RaynaldM — APPROVED** [review]

**Comment by raman-m** [comment]
@EngRajabi 
> @raman-m requested changes [on Jul 22](#pullrequestreview-1542095717)
> @wast reviewed [on Jul 31](#pullrequestreview-1555680140)

As a PR author, you have to work on code review issues!
Don't look at [Ray's approval](#pullrequestreview-1551519605). Sometime Ray approves PRs even without unit tests! And such PRs won't be merged until addition required unit tests.
Sometime PRs don't have documentation update, in case of new feature, or breaking changes. Such PRs won't be merged until updating required docs.
So, you need a final approval from me.
Let me know if you need a help, or in case of any questions.

### @RaynaldM  FYI

File: src/Ocelot/Cache/CacheKeyGenerator.cs
Comment (EngRajabi) [reply src/Ocelot/Cache/CacheKeyGenerator.cs]: resolve

**Comment by EngRajabi** [comment]
@raman-m 
add unit test for changing
correct text grammar (HEader)

**Comment by raman-m** [comment]
Ridiculous [changes and diff](https://github.com/ThreeMammals/Ocelot/compare/develop...EngRajabi:Ocelot:develop) in your develop branch!
Could you [Sync fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) for **develop** branch immediately please?

**Review by raman-m — APPROVED** [review]

**Comment by raman-m** [comment]
@EngRajabi @RaynaldM
"**To merge or not to merge?**", that is the question! 😄 

I did significant changes: refactoring, updated tests & docs. So, look at these changes finally.
And if you don't mind, I'm going to merge...

**Review by RaynaldM — APPROVED** [review]

**Comment by RaynaldM** [comment]
> @EngRajabi @RaynaldM "**To merge or not to merge?**", that is the question! 😄
> 
> I did significant changes: refactoring, updated tests & docs. So, look at these changes finally. And if you don't mind, I'm going to merge...

This kind of PR needs a beta package version no ?

**Review by EngRajabi — COMMENTED** [review]
approve.
thanks

**Comment by raman-m** [comment]
> @RaynaldM commented [on Oct 19](#issuecomment-1770109099)
>
> This kind of PR needs a beta package version no ?

As you know Ray, we have no time to release beta-versions. I don't see a benefit. And we don't have testers and no automated testing to check each feature.
It seems the code is OK now. I'm going to merge to develop branch, and the feature will be released in v21 (October's release).
If we will find some bug in this feature then I will do a hotfix release.

**Comment by raman-m** [comment]
@EngRajabi It was a long story, but your feature has been merged into develop as commit e92b103644a573a73e3f720b134b9baac86cf145.
Congrats! 🥳 

P.S. For future PRs it is better to discuss and create an issue first, and go with development if Ocelot team discussed.
Sometime we have something related in backlog, it is better to attach.

**Comment by ggnaegi** [comment]
Hello Mohsen, 

In the end, we didn't include your changes in the version yet because we were short of acceptance tests and, for my part, I wasn't sure I'd got what you wanted.

Of course, the functionality will be integrated, but we'll have to do a bit more work. 

I'm really sorry, I need to understand what the use case is. You would like to define a header key per route, and then, client-side, make the request with a defined value for your header key, right? If the value is defined, it has an impact on the generation of the MD5 hash. If you call endpoint 1, the GET method and the header value "test1", then afterwards with the value "test2", you'll have two different keys.

Okay, if I've understood your proposal correctly, I'd add a few changes:
- Why just one header key and no more? Other API gateways offer caching based on one or more header keys. Perhaps instead of "Header": "authorization", we could provide several such as "Headers":["authorization", "scopes", "test"].

There are other - long-standing - problems that I think could still be corrected, such as the use of the request body content to generate the MD5 hash. I think this should be **disabled by default** and only enabled if the admin really wants it.

```json
"FileCacheOptions": {
  "TtlSeconds": 15,
  "Region": "somename",
  "Headers": ["authorization", "test1", "test2"],
  "EnableRequestBodyHashing": false
}
```

As of the two virtual methods in DownstreamRequest, I think we should return the HttpContent object {get} only and think about an improved way of reading the content.

@raman-m, @RaynaldM  what do you think?

**Comment by EngRajabi** [comment]
> Hello Mohsen,
> 
> In the end, we didn't include your changes in the version yet because we were short of acceptance tests and, for my part, I wasn't sure I'd got what you wanted.
> 
> Of course, the functionality will be integrated, but we'll have to do a bit more work.
> 
> I'm really sorry, I need to understand what the use case is. You would like to define a header key per route, and then, client-side, make the request with a defined value for your header key, right? If the value is defined, it has an impact on the generation of the MD5 hash. If you call endpoint 1, the GET method and the header value "test1", then afterwards with the value "test2", you'll have two different keys.
> 
> Okay, if I've understood your proposal correctly, I'd add a few changes:
> 
> * Why just one header key and no more? Other API gateways offer caching based on one or more header keys. Perhaps instead of "Header": "authorization", we could provide several such as "Headers":["authorization", "scopes", "test"].
> 
> There are other - long-standing - problems that I think could still be corrected, such as the use of the request body content to generate the MD5 hash. I think this should be **disabled by default** and only enabled if the admin really wants it.
> 
> ```json
> "FileCacheOptions": {
>   "TtlSeconds": 15,
>   "Region": "somename",
>   "Headers": ["authorization", "test1", "test2"],
>   "EnableRequestBodyHashing": false
> }
> ```
> 
> As of the two virtual methods in DownstreamRequest, I think we should return the HttpContent object {get} only and think about an improved way of reading the content.
> 
> @raman-m, @RaynaldM what do you think?

Hello, thank you for your attention
My requirement is to have cache for each user.
For example, I want to cache a user's shopping list, the cache key should be for each user.
If we want to work on this more, yes, we can add more

**Comment by ggnaegi** [comment]
@EngRajabi Would you be willing to give me permissions on your forked repo so I could "help" or "try to help" a bit? Thanks!

**Comment by RaynaldM** [comment]
> Hello Mohsen,
> 
> In the end, we didn't include your changes in the version yet because we were short of acceptance tests and, for my part, I wasn't sure I'd got what you wanted.
> 
> Of course, the functionality will be integrated, but we'll have to do a bit more work.
> 
> I'm really sorry, I need to understand what the use case is. You would like to define a header key per route, and then, client-side, make the request with a defined value for your header key, right? If the value is defined, it has an impact on the generation of the MD5 hash. If you call endpoint 1, the GET method and the header value "test1", then afterwards with the value "test2", you'll have two different keys.
> 
> Okay, if I've understood your proposal correctly, I'd add a few changes:
> 
> * Why just one header key and no more? Other API gateways offer caching based on one or more header keys. Perhaps instead of "Header": "authorization", we could provide several such as "Headers":["authorization", "scopes", "test"].
> 
> There are other - long-standing - problems that I think could still be corrected, such as the use of the request body content to generate the MD5 hash. I think this should be **disabled by default** and only enabled if the admin really wants it.
> 
> ```json
> "FileCacheOptions": {
>   "TtlSeconds": 15,
>   "Region": "somename",
>   "Headers": ["authorization", "test1", "test2"],
>   "EnableRequestBodyHashing": false
> }
> ```
> 
> As of the two virtual methods in DownstreamRequest, I think we should return the HttpContent object {get} only and think about an improved way of reading the content.
> 
> @raman-m, @RaynaldM what do you think?

From a general point of view, and this applies to this issue, I think we need to stay focused on what Ocelot is all about: being an API Gateway.
There are as many special cases as there are use cases, and I think we should add only the minimum to Ocelot, but open it up, so that whoever has a special case can deal with it themselves.
For example, what I've done is to add :
 ```
public static IOcelotBuilder AddPolly<T>(this IOcelotBuilder builder,
        QosDelegatingHandlerDelegate delegatingHandler,
        Dictionary<Type, Func<Exception, Error>> errorMapping)
        where T : class, IPollyQoSProvider<HttpResponseMessage>
    {
        builder.Services
            .AddSingleton(errorMapping)
            .AddSingleton<IPollyQoSProvider<HttpResponseMessage>, T>()
            .AddSingleton(delegatingHandler);

        return builder;
    }

```
Now I can have my own PollyQoSProvider custom for my particular case.

Having Ocelot handle all possible uses, all my special cases, seems to me to be a bad idea for the product.

Let's keep Ocelot simple and easy to tune

**Comment by ggnaegi** [comment]
@RaynaldM I agree with you. But, here we have two problems: First, that the request's body content is hashed as soon as content is provided, that's not good. The second is that many API-Gateways offer caching based on headers. As of the flexibility, we are almost there since you can implement your own IOcelotCache and ICacheKeyGenerator. ```Services.TryAddSingleton<IOcelotCache<CachedResponse>, AspMemoryCache<CachedResponse>>()```
```Services.TryAddSingleton<ICacheKeyGenerator, CacheKeyGenerator>();```
But, you are write, maybe providing the Extension method for that purpose would be a great addition.

**Comment by raman-m** [comment]
> @ggnaegi commented [on Nov 28 at 8:43 am](#issuecomment-1829349202)

We must remove content body reading from default implementation by [Mr. Ali in 2019](https://github.com/ThreeMammals/Ocelot/blame/4476f8273ed83a7934b9705a6f56dfdb5b16a841/src/Ocelot/Cache/CacheKeyGenerator.cs#L11)! aka the [CacheKeyGenerator](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Cache/CacheKeyGenerator.cs) class!
We have to rename `CacheKeyGenerator` to `DefaultCacheKeyGenerator` and redesign the thing!
My and your Gui concern is request content reading for caching key construction, that's bad design by Mr. Ali!
Moreover, that's wrong to consider request body as a part of caching key because every request can have unique body which leads to invalid value in cache, which forces to expiring of cache value. Finally I would say in 95% of cases even requests with GET verb will have new key and cache will be renewed each time. That's terrible design because performance degrades! Technically this bug switches off the caching in ~50% of requests. I am a bit shocked!... 😮😞 This sis my vision and understanding of this problem...

Default cache key generator must be based on URL and verbs only! ☝️ 
So, [these lines](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Cache/CacheKeyGenerator.cs#L10-L14) must be removed at all:
```csharp
if (downstreamRequest.Content != null)
{
    var requestContentString = Task.Run(async () => await downstreamRequest.Content.ReadAsStringAsync()).Result;
    downStreamUrlKeyBuilder.Append(requestContentString);
}
```
Do you agree?

Regarding the [ICacheKeyGenerator](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/DependencyInjection/OcelotBuilder.cs#L119) interface injection... Here is screenshot of current references:
image

So, The interface is already a part of DI and its implementation can be overridden by custom user's version.
I don't understand Mohsen's request. His team can override this interface and a couple of related ones, and enjoy the new caching thing... But he requested to merge his feature to Ocelot... This is rare case, and sure it could be reused by the community...
Anyway, his PR & feature should be redesigned, enhanced and team should have 100% consensus to merge.
Sure, we will help him...
But first we have to redesign and resolve "content reading" logic problem.
Agree?

**Comment by raman-m** [comment]
@RaynaldM commented [on Nov 28 at 7:55 am](#issuecomment-1829283501):

Hello, Raynald!
Thanks for being online!

----
> I think it's a bad idea to make Ocelot handle all possible uses and all special cases.

Agree partially! This is special rare user case, but... We can abstract the logic, ensure top performance in caching, and then introduce cache control and management headers. That is Mohsen's idea... Also, as @ggnaegi said, a lot of gateways have caching feature based on headers. I believe this is the best practice to manage cache by custom headers.

----
>Let's keep Ocelot simple and easy to tune

My understanding... Agree, we need to provide common caching and ability to switch on/off the feature. To be fair, we have the c current Caching feature... This solution should be lite and functioning. Also, we must decouple everything, all related interfaces should go to DI container with injected default implementations with lite design.
That's how we can allow to developers to extend the Ocelot pipeline.

**Comment by raman-m** [comment]
@EngRajabi So, your feature is right: don't worry! 😉 But we've detected some super important things in related logic to work on before merging and delivering your feature. I've marked this PR adding to [Nov'23](https://github.com/ThreeMammals/Ocelot/milestone/2) milestone, but I'm not sure we have time to work on the feature in November... So, I will add new PR to [Dec'23](https://github.com/ThreeMammals/Ocelot/milestone/3) release if you will be pro-active... 🆗 ❔ 

Finally, could you open new PR from this feature branch, plz?
But before opening new
- Add @ggnaegi and/or me as a collaborator to your forked repo please. 
- Make develop branch identical to [ThreeMammals:develop](https://github.com/ThreeMammals/Ocelot/tree/develop)
- Rebase feature branch onto develop one
- And create new PR please

Good luck!

P.S. Creation a new PR by you is essential because you will be author of the feature 😉 
I don't want to create feature branch in head repo to keep your authority... 😄 

@ggnaegi  You are the mentor for this feature. Okay?

**Comment by ggnaegi** [comment]
@raman-m ok, fine for me

**Comment by raman-m** [comment]
@EngRajabi @ggnaegi 
Good plan! Let's work on it! 😉

**Comment by ggnaegi** [comment]
@RaynaldM @raman-m @EngRajabi To summarize your wishes:
- The Cache DI should be refactored. It shouldn't be part of main Ocelot Builder
- We should provide Extension methods to use the default cache or the users can DI their own caching mechanisms eg. 
```c#
s.AddOcelot()
  .AddCache();
```
or
```c#
s.AddOcelot()
.AddCache<MyImplementation>();
```
- In the default implementation, we would include what I suggested to ensure backward compatibility.
```json
"FileCacheOptions": {
  "TtlSeconds": 15,
  "Region": "somename",
  "Headers": ["authorization", "test1", "test2"],
  "EnableRequestBodyHashing": false
}
```
- What should we do with ```Ocelot.Cache.CacheManager``` ?

**Comment by ggnaegi** [comment]
@EngRajabi, @raman-m  I came up with this idea... It would fit the needs for the following FileCachOptions 
```json
"FileCacheOptions": {
  "TtlSeconds": 15,
  "Region": "somename",
  "Headers": ["authorization", "test1", "test2"],
  "EnableRequestBodyHashing": false
}
```

```c#
using Ocelot.Configuration;
using Ocelot.Request.Middleware;

namespace Ocelot.Cache;

public class CacheKeyGenerator : ICacheKeyGenerator
{
    private const char Delimiter = '-';
    private readonly IMemoryStreamManager _memoryStreamManager;
    private readonly Encoding _encoding = new UTF8Encoding(false);

    public CacheKeyGenerator(IMemoryStreamManager memoryStreamManager)
    {
        _memoryStreamManager = memoryStreamManager;
    }

    public async ValueTask<string> GenerateRequestCacheKey(DownstreamRequest downstreamRequest,
        DownstreamRoute downstreamRoute)
    {
        using var memoryStream = _memoryStreamManager.GetStream();
        await using var writer = new StreamWriter(memoryStream, _encoding);

        await writer.WriteAsync(downstreamRequest.Method);
        await writer.WriteAsync(Delimiter);
        await writer.WriteAsync(downstreamRequest.OriginalString);

        var cacheOptions = downstreamRoute.CacheOptions;
        if (cacheOptions == null)
        {
            return await GenerateMd5(writer, memoryStream);
        }

        await AppendHeadersValues(writer, cacheOptions.Headers, downstreamRequest);
        var flush = await AppendHashedBodyContent(writer, cacheOptions.RequestBodyHashing, downstreamRequest);
        return await GenerateMd5(writer, memoryStream, flush);
    }

    private static async Task<string> GenerateMd5(StreamWriter writer, MemoryStream memoryStream, bool flush = true)
    {
        if (flush)
        {
            await writer.FlushAsync();
        }

        return MD5Helper.GenerateMd5(memoryStream);
    }

    private static async Task AppendHeadersValues(StreamWriter writer, IReadOnlyCollection<string> headers,
        DownstreamRequest downstreamRequest)
    {
        if (headers == null || headers.Count == 0)
        {
            return;
        }

        foreach (var headerKey in headers)
        {
            string header = null;
            if (downstreamRequest.Headers.Contains(headerKey))
            {
                header = downstreamRequest.Headers.GetValues(headerKey).FirstOrDefault();
            }

            if (string.IsNullOrEmpty(header))
            {
                continue;
            }

            await writer.WriteAsync(Delimiter);
            await writer.WriteAsync(header);
        }
    }

    private static async Task<bool> AppendHashedBodyContent(StreamWriter writer, bool requestBodyHashing,
        DownstreamRequest downstreamRequest)
    {
        if (!requestBodyHashing || downstreamRequest.Content == null)
        {
            return true;
        }

        var contentBytes = await downstreamRequest.Content.ReadAsByteArrayAsync();
        if (contentBytes.Length == 0)
        {
            return true;
        }

        await writer.WriteAsync(Delimiter);
        await writer.FlushAsync(); // Ensure the delimiter is written to the base stream

        // Write the byte array directly to the base stream
        await writer.BaseStream.WriteAsync(contentBytes);
        return false;
    }
}
```

```c#
using Microsoft.IO;

namespace Ocelot.Cache;

public class MemoryStreamManager : IMemoryStreamManager
{
    private readonly RecyclableMemoryStreamManager _recyclableMemoryStreamManager;

    public MemoryStreamManager()
    {
        _recyclableMemoryStreamManager = new RecyclableMemoryStreamManager();
    }

    public MemoryStream GetStream()
    {
        return _recyclableMemoryStreamManager.GetStream();
    }
}
```

**Comment by raman-m** [comment]
@ggnaegi commented [on Nov 28](#issuecomment-1830740495):
> I came up with this idea...

If this is idea, why not to create discussion first in [Ideas](https://github.com/ThreeMammals/Ocelot/discussions/categories/ideas) category?

Please move all your comments to a new discussion thread!
I don't want to keep conversation in this closed PR...

**Comment by raman-m** [comment]
@EngRajabi  Mohsen, where are you?

Why not to follow [this plan](https://github.com/ThreeMammals/Ocelot/pull/1172#issuecomment-1829642165)?
**Could you create new PR ASAP please** ❓
We will include it to current release...

**Comment by EngRajabi** [comment]
> @EngRajabi Mohsen, where are you?
> 
> Why not to follow [this plan](https://github.com/ThreeMammals/Ocelot/pull/1172#issuecomment-1829642165)? **Could you create new PR ASAP please** ❓ We will include it in current upcoming [Nov'23](https://github.com/ThreeMammals/Ocelot/milestone/2) release...

Hello
I followed the ideas, they are very good. But maybe I can't implement the code myself

**Comment by raman-m** [comment]
@EngRajabi
What we've discussed here by Ocelot Core team will be moved to separate feature on the top of your one. You will not develop it.
But... We need to deliver your old feature to develop once again!
You need to create new PR from the current feature branch, and we go to deliver...
As a PR author you will be the author of the feature. You see now?
If you don't want to create PR then we can manage by ourself.

**Comment by EngRajabi** [comment]
@raman-m 
Hello, please create it yourself.

**Comment by raman-m** [comment]
> @EngRajabi commented [on Dec 9](https://github.com/ThreeMammals/Ocelot/pull/1172#issuecomment-1848245818)

Hey, Mohsen!
Your feature has been delivered once again and finally by the commit https://github.com/ThreeMammals/Ocelot/commit/ba641b245968294245d487a7c33d1bcc039f789c!
Congrats! :partying_face:

**Comment by EngRajabi** [comment]
> > @EngRajabi commented [on Dec 9](https://github.com/ThreeMammals/Ocelot/pull/1172#issuecomment-1848245818)
> 
> Hey, Mohsen! Your feature has been delivered once again and finally by the commit [ba641b2](https://github.com/ThreeMammals/Ocelot/commit/ba641b245968294245d487a7c33d1bcc039f789c)! The feature will be released soon: it is included in current upcoming Nov'23 release. Congrats! 🥳

thanks

**Pull request merged**
